File: Python_Projects/100_Game.py
# Purpose: The game is designed to take input from two players (First Player and Second Player)
# it adds numbers from both sides to total_sum till one of them adds a number that makes the total_sum be 100 or more
def game_100():
    # Started with declaring the variables to make them in a bigger scope.
    first_player = 0
    second_player = 0
    total_sum = int(0)
    while total_sum < 100:
        # Using try and except to handle the error of entering a non integer number.
        try:
            print(f"current Total Sum is: {total_sum}")
            # If an error occurred in the first or second player input we will restart the round.
            # When restarting the round we need to make sure that we removed what happened in it.
            first_is_added = 0
            min_num = min(10, (100 - total_sum))
            first_player = int(input(f"First player enter a number from 1 to {min_num}: "))
            # next while loop will only work when the input from First Player is less than 1 or
            # more than minimum of 10 and (100 - total_sum)
            while not 1 <= first_player <= min_num:
                print("Invalid input entered from first player ", end='')
                print(f"make sure to input a number from 1 to {min_num}.")
                first_player = int(input(f"First player enter a number from 1 to {min_num}: "))
            total_sum += first_player
            first_is_added = 1
            # the next condition is to know if the First Player won after adding his input to the total_sum or not.
            if total_sum == 100:
                print("Congratulations! First player won.")
                break
            min_num = min(10, (100 - total_sum))
            print(f"current Total Sum is: {total_sum}")
            second_player = int(input(f"Second player enter a number from 1 to {min_num}: "))
            # next while loop will only work when the input from Second Player is less than 1 or
            # more than minimum of 10 and (100 - total_sum)
            while not 1 <= second_player <= min_num:
                print("Invalid input entered from second player make ", end='')
                print(f"sure to input a number from 1 to {min_num}.")
                second_player = int(input(f"Second player enter a number from 1 to {min_num}: "))
            total_sum += second_player
            # the next condition is to know if the Second Player won after adding his input to the total_sum or not.
            if total_sum == 100:
                print("Congratulations! Second player won.")
                break
        # We make the exception in error to print it to the Players.
        except Exception as error:
            # If first_is_added equal 1 then that means that the error is from The Second Player.
            # Which means that we need to remove The First Player's input from the total_sum.
            # If first_is_added equal 0 we continue because the error occurred from The First Player.
            if first_is_added == 1:
                total_sum -= first_player
            print(f"Invalid input {error} entered from a player. restarting the round.")


print("This is 100 game.")
print("The game is designed to take input from two players (First Player and Second Player)", end='')
print("it adds numbers from both sides to total_sum till one of ", end='')
print("them adds a number that makes the total_sum equal 100 and that means that he won")
print("_______________________________________________________________________________________________________________")
print("")
while 1:
    # I'm using .lower() built-in function to handle uppercased inputs (e.g. sTart, STart, START, ...etc)
    choice = input("Welcome in 100GAME. Type 'Start' to start a new game. or type 'Quit' to quit. \n").lower()
    if choice == "start":
        game_100()
    # If The Players entered 'quit' we terminate the program.
    elif choice == "quit":
        break
    # This else is to handle the case that the user enters something else other than start or quit.
    else:
        print(f"'{choice}' Is an invalid choice. Enter one of the two choices.")
        continue

File: C++_projects/100_Game.cpp
/*
// Program: This is 100 game.
            The game is designed to take input from two players (First Player and Second Player)
            it adds numbers from both sides to total_sum till one of them adds a number that
            makes the total_sum be 100.
// Date: 21/2/2024 "I made Version 1 of it after doing it in python directly"
// Last_Update: 1/3/2024 "I made it better by using integer 'mini' and some other features in the code"
// Version: 2.0
 */
#include <bits/stdc++.h>
using namespace std;
int main(){
    cout << "This is 100 game.\n";
    cout << "The game is designed to take input from two players (First Player and Second Player)\n";
    cout << "The game adds numbers from both sides to total_sum till one of ";
    cout << "them adds a number that makes the total_sum equal 100 and that means that he won\n";
    cout << "_______________________________________________________________________________________________\n";
    while(true){
        string choice;
        cout << "Welcome in 100GAME. Type 'Start' to start a new game. or type 'Quit' to quit.\n";
        cin >> choice;
        int first_player, second_player, total_sum = 0, mini = INT_MAX;
        // Next for loop to change the input to lower case using tolower to change each character.
        for (auto &c : choice){
            c = tolower(c);
        }
        if(choice == "start"){
            // While the total_sum doesn't exceed 100 the game continues
            while(total_sum < 100){
                cout << "Current total_sum is: " << total_sum << '\n';
                // 'mini' Is used to make the user enter a number from 1 to mini which is 10 or lower.
                mini = min(10, (100 - total_sum));
                cout << "First player enter a number from 1 to " << mini << '\n';
                cin >> first_player;
                /*
                 * Next for loop for wrong input and using cin.fail(), cin,clear(),
                 * cin.ignore(numeric_limits<streamsize>::max(), '\n') to remove all wrong input from the buffer that
                 * holds the input.
                 */
                while (cin.fail() || first_player < 1 || first_player > mini) {
                    cin.clear();
                    cin.ignore(numeric_limits<streamsize>::max(), '\n');
                    cout << "Invalid input entered from first player make sure to input a number from 1 to " << mini << '\n';
                    cout << "First player enter a number from 1 to " << mini << '\n';
                    cin >> first_player;
                }
                // Updating total_sum.
                total_sum += first_player;
                // Checking if the First Player won or not.
                if (total_sum == 100){
                    cout << "Congratulations! First player won." << endl;
                    break;
                }
                mini = min(10, (100 - total_sum));
                cout << "Current total_sum is: " << total_sum << '\n';
                cout << "Second player enter a number from 1 to " << mini << '\n';
                cin >> second_player;
                /*
                 * Next for loop for wrong input and using cin.fail(), cin,clear(),
                 * cin.ignore(numeric_limits<streamsize>::max(), '\n') to remove all wrong input from the buffer that
                 * holds the input.
                 */
                while (cin.fail() || second_player < 1 || second_player > mini){
                    cin.clear();
                    cin.ignore(numeric_limits<streamsize>::max(), '\n');
                    cout << "Invalid input entered from Second player make sure to input a number from 1 to " << mini << '\n';
                    cout << "Second player enter a number from 1 to " << mini << '\n';
                    cin >> second_player;
                }
                // Updating total_sum.
                total_sum += second_player;
                // Checking if the Second Player won or not.
                if (total_sum == 100){
                    cout << "Congratulations! Second player won." << endl;
                    break;
                }
            }
        }
        // If The Players entered 'quit' we terminate the program.
        else if(choice == "quit"){
            break;
        }
        // This else is to handle the case that the user enters something else other than start or quit.
        else{
            cout << '{' << choice << "} Is an invalid choice. Enter one of the two choices.\n";
            continue;
        }
    }
    return 0;
}
